pcsx2/SPU2/SndOut_Cubeb.cpp

```diff
@@ -0,0 +1,341 @@
+#ifdef _WIN32
+		HRESULT hr = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+		m_COMInitializedByUs = SUCCEEDED(hr);
+		if (FAILED(hr) && hr != RPC_E_CHANGED_MODE)
+		{
+			Console.Error("Failed to initialize COM");
+			return false;
+		}
+#endif
+
+		int rv = cubeb_init(&m_context, "PCSX2", m_Backend.empty() ? nullptr : m_Backend.c_str());
+		if (rv != CUBEB_OK)
+		{
+			Console.Error("Could not initialize cubeb context: %d", rv);
+			return false;
+		}
+
+		switch (numSpeakers) // speakers = (numSpeakers + 1) *2; ?
+		{
+			case 0:
+				channels = 2;
+				break; // Stereo
+			case 1:
+				channels = 4;
+				break; // Quadrafonic
+			case 2:
+				channels = 6;
+				break; // Surround 5.1
+			case 3:
+				channels = 8;
+				break; // Surround 7.1
+			default:
+				channels = 2;
+				break;
+		}
+
+		switch (channels)
+		{
+			case 2:
+				Console.WriteLn("* SPU2 > Using normal 2 speaker stereo output.");
+				ActualReader = std::make_unique<ConvertedSampleReader<StereoOut16>>(&writtenSoFar);
+				break;
+
+			case 3:
+				Console.WriteLn("* SPU2 > 2.1 speaker expansion enabled.");
+				ActualReader = std::make_unique<ConvertedSampleReader<Stereo21Out16>>(&writtenSoFar);
+				break;
+
+			case 4:
+				Console.WriteLn("* SPU2 > 4 speaker expansion enabled [quadraphenia]");
+				ActualReader = std::make_unique<ConvertedSampleReader<Stereo40Out16>>(&writtenSoFar);
+				break;
+
+			case 5:
+				Console.WriteLn("* SPU2 > 4.1 speaker expansion enabled.");
+				ActualReader = std::make_unique<ConvertedSampleReader<Stereo41Out16>>(&writtenSoFar);
+				break;
+
+			case 6:
+			case 7:
+				switch (dplLevel)
+				{
+					case 0:
+						Console.WriteLn("* SPU2 > 5.1 speaker expansion enabled.");
+						ActualReader = std::make_unique<ConvertedSampleReader<Stereo51Out16>>(&writtenSoFar); //"normal" stereo upmix
+						break;
+					case 1:
+						Console.WriteLn("* SPU2 > 5.1 speaker expansion with basic ProLogic dematrixing enabled.");
+						ActualReader = std::make_unique<ConvertedSampleReader<Stereo51Out16Dpl>>(&writtenSoFar); // basic Dpl decoder without rear stereo balancing
+						break;
+					case 2:
+						Console.WriteLn("* SPU2 > 5.1 speaker expansion with experimental ProLogicII dematrixing enabled.");
+						ActualReader = std::make_unique<ConvertedSampleReader<Stereo51Out16DplII>>(&writtenSoFar); //gigas PLII
+						break;
+				}
+				channels = 6; // we do not support 7.0 or 6.2 configurations, downgrade to 5.1
+				break;
+
+			default: // anything 8 or more gets the 7.1 treatment!
+				Console.WriteLn("* SPU2 > 7.1 speaker expansion enabled.");
+				ActualReader = std::make_unique<ConvertedSampleReader<Stereo71Out16>>(&writtenSoFar);
+				channels = 8; // we do not support 7.2 or more, downgrade to 7.1
+				break;
+		}
+
+		cubeb_stream_params params = {};
+		params.format = CUBEB_SAMPLE_S16LE;
+		params.rate = SampleRate;
+		params.channels = channels;
+		params.layout = CUBEB_LAYOUT_UNDEFINED;
+		params.prefs = CUBEB_STREAM_PREF_PERSIST;
```

Comment:
In cubeb.h, it mentions that CUBEB_STREAM_PREF_PERSIST is obsolete and ignored by all backends. We could probably just use CUBEB_STREAM_PREF_NONE instead...